clvm-traits/src/from_clvm.rs

```diff
@@ -1,47 +1,70 @@
-use std::array::TryFromSliceError;
+use clvmr::{allocator::NodePtr, Allocator};
+use num_bigint::{BigInt, Sign};
 
-use clvmr::{
-    allocator::{NodePtr, SExp},
-    op_utils::nullp,
-    Allocator,
-};
-use num_bigint::Sign;
+use crate::{ClvmDecoder, FromClvmError};
 
-use crate::{Error, Result};
+pub trait FromClvm<N>: Sized {
+    fn from_clvm(decoder: &impl ClvmDecoder<Node = N>, node: N) -> Result<Self, FromClvmError>;
+}
+
+pub trait FromNodePtr {
+    fn from_node_ptr(a: &Allocator, node: NodePtr) -> Result<Self, FromClvmError>
+    where
+        Self: Sized;
+}
 
-pub trait FromClvm: Sized {
-    fn from_clvm(a: &Allocator, ptr: NodePtr) -> Result<Self>;
+impl<T> FromNodePtr for T
+where
+    T: FromClvm<NodePtr>,
+{
+    fn from_node_ptr(a: &Allocator, node: NodePtr) -> Result<Self, FromClvmError>
+    where
+        Self: Sized,
+    {
+        T::from_clvm(a, node)
+    }
 }
 
-impl FromClvm for NodePtr {
-    fn from_clvm(_a: &Allocator, ptr: NodePtr) -> Result<Self> {
-        Ok(ptr)
+impl FromClvm<NodePtr> for NodePtr {
+    fn from_clvm(
+        _decoder: &impl ClvmDecoder<Node = NodePtr>,
+        node: NodePtr,
+    ) -> Result<Self, FromClvmError> {
+        Ok(node)
     }
 }
 
 macro_rules! clvm_primitive {
     ($primitive:ty) => {
-        impl FromClvm for $primitive {
-            fn from_clvm(a: &Allocator, ptr: NodePtr) -> Result<Self> {
-                if let SExp::Atom = a.sexp(ptr) {
-                    let (sign, mut vec) = a.number(ptr).to_bytes_be();
-                    if vec.len() < std::mem::size_of::<$primitive>() {
-                        let mut zeros = vec![0; std::mem::size_of::<$primitive>() - vec.len()];
-                        zeros.extend(vec);
-                        vec = zeros;
-                    }
-                    let value =
-                        <$primitive>::from_be_bytes(vec.as_slice().try_into().map_err(
-                            |error: TryFromSliceError| Error::Custom(error.to_string()),
-                        )?);
-                    Ok(if sign == Sign::Minus {
-                        value.wrapping_neg()
-                    } else {
-                        value
-                    })
-                } else {
-                    Err(Error::ExpectedAtom(ptr))
+        impl<N> FromClvm<N> for $primitive {
+            fn from_clvm(
+                decoder: &impl ClvmDecoder<Node = N>,
+                node: N,
+            ) -> Result<Self, FromClvmError> {
+                const LEN: usize = std::mem::size_of::<$primitive>();
+
+                let bytes = decoder.decode_atom(&node)?;
+                let number = BigInt::from_signed_bytes_be(bytes);
```

Comment:
Yeah I'll do an optimization pass later, I just don't want to rush it since it's a pretty critical bit of code - last time I touched it I ended up with numbers being encoded wrong in the wallet I'm building